### ISSUE TITLE
Make unsafe: true the default setting for goldmark(?)

### ISSUE THREAD

**Issue by Brixy** [gohugoio/hugo#6581]
This is opiniated because I haven’t given much thought to the drawbacks yet:

In the forums we have seen many questions about Hugo 0.60.n omitting HTML in markdown files. The reason for this is clear and well documented.

As HTML blocks are *allowed* both in Markdown and [CommonMark](https://spec.commonmark.org/0.29/#html-block) I personally think it would be better to use `unsafe: true` as the default. Otherwise the default behaviour feels kind of non-compliant with CommonMark.

**Comment by bep** [comment]
Have you considered the security aspects of your proposal?

**Comment by Brixy** [comment]
Thanks for your time and comment.

I know that this is the reason for your design decision.

Explaining potential security issues and giving the option to use safe mode would be the opposite way. If that is just too dangerous we can close this issue.

**Comment by bep** [comment]
>Explaining potential security issues and giving the option to use safe mode would be the opposite way.

I have thought about this to some extent. The thing is ...

* People do not check the default values unless it breaks something visible
* For some pretty HTML table that's no longer there, the above is annoying, I agree.
* If you get injected some malicious code in your site, you wouldn't know until some user complains that his/her bank account is empty.

I understand that the above security example is far fetched, esp. for people in full control over their content files. But the thing is, the average Hugo user does not care about "other Hugo users' security". And the only way to get peoples' attention about this is to have every (sensible) security option turned on by default.

Note that I have plans to add Bluemonday as an option (which may disrupt a little what I said above...), which is a less restrictive HTML sanitizer. But time is of the essence.

**Comment by kaushalmodi** [comment]
> If you get injected some malicious code in your site, you wouldn't know until some user complains that his/her bank account is empty.

Setting `unsafe = false` will not prevent that.

The option prevents use of HTML in the Markdown content only at the time of the site generation, at the point where the site owner is in complete control. 

The malicious injection happens once the site is already deployed. 

To prevent such injections, the user should look into proper solutions like [Content Security Policy][1]. 

The `unsafe=false` is just a red herring. Truly unsafe things can happen using `getJSON` that we already have. 

I'd vote to set unsafe default to true as well. If user is truly wanting to secure their site, they much switch to https + use CSP. 

[1]: https://developers.google.com/web/fundamentals/security/csp

**Comment by bep** [comment]
>The option prevents use of HTML in the Markdown content only at the time of the site generation, at the point where the site owner is in complete control.

You are wrong.

I don't have time make this into a ping pong debate, so I'm closing this. But while you are right that the "site owner is in complete control", if that site owner edits content contributed from millions of people on the interweb (via PRs and similar), it's pretty easy to envision that that editor will let some unsafe content slip through and deployed.

**Comment by kaushalmodi** [comment]
It would be nice if you gave a real example of how what I said was untrue. 

I don't understand why Hugo needs to be a security-nanny here (when this option just gives a false sense of security) when the Markdown and CommonMark do allow HTML in Markdown. 

About your comment about PR's introducing malware, the site owner needs to be educated enough to prevent there. There are no set of "safety variables" that can overcome that. 

What if someone PR'd a malicious JS to a Hugo template? Is it Hugo's job to prevent that too?

Bottom line: That option introduces more inconvenience than any security.  There's no real example of malicious code injection (other than a malicious PR .. that too only in Markdown content?) that the current default protects against.

**Comment by bep** [comment]
>It would be nice if you gave a real example of how what I said was untrue.

I wasn't going to get lured into a debate about this, but as I assume people will read this thread later, I will post a final comment.

For content, variations of this are what the `Unsafe=true` would allow:

```go
package main

import (
	"bytes"
	"fmt"
	"log"

	"github.com/yuin/goldmark/renderer/html"

	"github.com/yuin/goldmark"
)

func main() {
	unsafeContent := `
<script>
alert("XSS");
</script>
`
	markdown := goldmark.New(
		goldmark.WithRendererOptions(
			html.WithUnsafe(),
		),
	)
	var buf bytes.Buffer
	err := markdown.Convert([]byte(unsafeContent), &buf)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(buf.String())
}

```

* For HTML templates, this is the basic security model: https://golang.org/pkg/html/template/#hdr-Security_Model
* Sure, you can get insecure PRs, but it's much easier to audit 10 templates than 10 million content files. You can write the templates yourself and refuse any PR.

So, security is hard and Hugo certainly does not come with an "OWAS guaranteed secure" label for every site we generate. But we at least have a security mindset/model. This is why we use Go's `html/template` package even if it's painful, this is why `Unsecure=false` by default, this is why we have a carefully sandboxed and read-only filesystem with symbolic link support only for the main project. To point at some other random area with some potential security risk and argue that that makes every other security remedy useless, that is an argument I'm not buying.

I repeat;

Hugo's security may not be perfect. None are. It's easy enough to turn off those security features if you trust your content/templates/data. But we cannot have them default off.

**Comment by kaushalmodi** [comment]
I am not denying that example code you gave.. you are injecting a script tag.. 

An astute site owner should spot such when merging PR's.. (who receives million line  PR's for Markdown btw? :)).

What I said is still true though.. the unsafe=false "security" stops once the user generates the site using Hugo. 

Once the site is generated, everything is HTML! (This is for all readers, not you)

**Comment by bep** [comment]
@kaushalmodi I think we can agree to disagree on this.

**Comment by kaushalmodi** [comment]
@bep Here's another take..

Let's say the user has unsafe=false and that user unknowingly merged a PR with a malicious script tag in the Markdown content. 

As expected, that script won't run.

But that malicious script is still in the user's Markdown content!!

What if that unsafe=false is later changed to unsafe=true later for the necessity to allow valid HTML tags like class and div. 

The best way is that one just be responsible when merging PR's. 

/cc @budparr

**Comment by bep** [comment]
>The best way is that one just be responsible when merging PR's.

I suggest both.

**Comment by bep** [comment]
>An astute site owner should spot such when merging PR's.. (who receives million line PR's for Markdown btw? :)).

I have seen full translations of the Lets Encrypt website/docs come in the form a PR. And when it's in some foreign lingo it may be easy to spot some random script deep inside.
